src/System.Collections/src/System/Collections/Generic/Queue.cs

```diff
@@ -234,15 +234,18 @@ IEnumerator IEnumerable.GetEnumerator()
         // InvalidOperationException.
         public T Dequeue()
         {
-            if (_size == 0)
+            int head = _head;
+            T[] array = _array;
```

Comment:
Are the array copies here necessary?

Reply:
It may not be necessary but the JIT is bugged and not having this copy may result in incorrect range check elimination. See the discussion in the similar `Stack` PR: https://github.com/dotnet/corefx/pull/26086#discussion_r159078422